File: packages/node/src/sinks/postgres.ts
import type { Sink } from './types.js';
import type { AnyEvent, ToolCallEvent } from '../events.js';

export interface PostgresSinkOptions {
  /** An existing `pg` Pool to reuse. If omitted, a new Pool is created from `pg`'s own env var defaults (PGHOST, PGUSER, PGPASSWORD, PGDATABASE, PGPORT, or a `connectionString`). */
  pool?: unknown;
  connectionString?: string;
  toolCallTable?: string;
}

interface QueryablePool {
  query: (text: string, values: unknown[]) => Promise<unknown>;
}

const TOOL_CALL_COLUMNS = [
  'ts',
  'server_name',
  'server_version',
  'tool_name',
  'session_id',
  'agent_id',
  'client_name',
  'client_version',
  'user_id',
  'org_id',
  'duration_ms',
  'success',
  'error_kind',
  'error_message',
  'request_bytes',
  'response_bytes',
  'arguments',
  'intent',
  'transport'
] as const;

function toolCallRow(event: ToolCallEvent): unknown[] {
  return TOOL_CALL_COLUMNS.map(column => event[column]);
}

/**
 * Builds one `insert into <table> (cols) values ($1,...),($n+1,...),...`
 * statement for all rows, with a flat bind array in column order. `pg`
 * encodes each bound value itself (Date -> timestamp string, plain object ->
 * JSON), so rows are passed through as-is.
 *
 * Postgres caps a statement at 65535 bind parameters; at 19 columns that is
 * 3449 rows per statement, far above the default buffer size of 20, so the
 * rows are not chunked.
 */
function buildMultiRowInsert(
  table: string,
  columns: readonly string[],
  rows: unknown[][]
): { text: string; values: unknown[] } {
  const values: unknown[] = [];
  const tuples: string[] = [];
  for (const row of rows) {
    const placeholders = row.map((_, i) => `$${values.length + i + 1}`);
    tuples.push(`(${placeholders.join(',')})`);
    values.push(...row);
  }
  const text = `insert into ${table} (${columns.join(', ')}) values ${tuples.join(',')}`;
  return { text, values };
}

/**
 * Writes rows into the table defined by schema/events.md's Postgres DDL.
 * Requires the optional peer dependency `pg` — dynamically imported so it
 * isn't required unless this sink is actually used.
 *
 * Each `write()` issues at most one multi-row insert for the batch.
 */
export function postgresSink(options: PostgresSinkOptions = {}): Sink {
  const toolCallTable = options.toolCallTable ?? 'mcpsignals_tool_call';

  let poolPromise: Promise<QueryablePool> | undefined;

  async function getPool() {
    if (!poolPromise) {
      poolPromise = (async () => {
        if (options.pool) {
          return options.pool as QueryablePool;
        }
        const pg = await import('pg');
        const Pool = pg.Pool ?? pg.default.Pool;
        return new Pool(
          options.connectionString ? { connectionString: options.connectionString } : undefined
        );
      })();
    }
    return poolPromise;
  }

  return {
    async write(events: AnyEvent[]): Promise<void> {
      const toolCallRows = events
        .filter(event => event.event_type === 'tool_call')
        .map(event => toolCallRow(event));
      if (toolCallRows.length === 0) {
        return;
      }

      const pool = await getPool();
      const { text, values } = buildMultiRowInsert(toolCallTable, TOOL_CALL_COLUMNS, toolCallRows);
      await pool.query(text, values);
    }
  };
}

File: packages/python/src/mcpsignals/sinks/postgres.py
"""Postgres sink. Requires the `postgres` extra (asyncpg). Table DDL lives in
schema/events.md. Connection info comes from asyncpg's own env var defaults
(PGHOST, PGPORT, PGUSER, PGPASSWORD, PGDATABASE) or an explicit dsn/pool -
never an invented config file.
"""

from mcpsignals.events import ToolCallEvent

_TOOL_CALL_COLUMNS = [
    "ts",
    "server_name",
    "server_version",
    "tool_name",
    "session_id",
    "agent_id",
    "client_name",
    "client_version",
    "user_id",
    "org_id",
    "duration_ms",
    "success",
    "error_kind",
    "error_message",
    "request_bytes",
    "response_bytes",
    "arguments",
    "intent",
    "transport",
    "protocol_version",
    "request_id",
    "trace_id",
    "parent_span_id",
    "result_type",
    "error_code",
    "read_only_hint",
    "destructive_hint",
]


class PostgresSink:
    def __init__(self, dsn: str | None = None, pool=None):
        """Provide either `dsn` (asyncpg will connect lazily) or an existing
        `pool` (an asyncpg.Pool you already manage). If neither is given,
        asyncpg's own environment-variable defaults apply.
        """
        self._dsn = dsn
        self._pool = pool

    async def _get_pool(self):
        if self._pool is None:
            # Local import, and only on the path that actually needs it: a
            # caller who injected their own pool never requires the `postgres`
            # extra to be installed at all.
            import asyncpg

            self._pool = await asyncpg.create_pool(dsn=self._dsn)
        return self._pool

    async def write(self, events: list[ToolCallEvent]) -> None:
        import json as _json

        tool_calls = [e for e in events if isinstance(e, ToolCallEvent)]
        if not tool_calls:
            # Nothing to write, so never open a connection (or a pool) for it.
            return

        pool = await self._get_pool()
        async with pool.acquire() as conn:
            if tool_calls:
                rows = [
                    (
                        e.ts,
                        e.server_name,
                        e.server_version,
                        e.tool_name,
                        e.session_id,
                        e.agent_id,
                        e.client_name,
                        e.client_version,
                        e.user_id,
                        e.org_id,
                        e.duration_ms,
                        e.success,
                        e.error_kind,
                        e.error_message,
                        e.request_bytes,
                        e.response_bytes,
                        _json.dumps(e.arguments) if e.arguments is not None else None,
                        e.intent,
                        e.transport,
                        e.protocol_version,
                        e.request_id,
                        e.trace_id,
                        e.parent_span_id,
                        e.result_type,
                        e.error_code,
                        e.read_only_hint,
                        e.destructive_hint,
                    )
                    for e in tool_calls
                ]
                await conn.copy_records_to_table(
                    "mcpsignals_tool_call", records=rows, columns=_TOOL_CALL_COLUMNS
                )
